### ISSUE TITLE
[Question] Validating Sometimes, Implicit Required Conditionals and Primitives

### ISSUE THREAD

**Issue by derekmd** [laravel/internals#472]
## Optional numeric `<input>` form

Here's a screenshot of a simple form row of three numeric `<input>` fields

![form](https://cloud.githubusercontent.com/assets/823566/23977189/00df7194-09c3-11e7-8648-65814136d32d.png)

Markup:

```html
<div class="panel">
    <div class="row">
        <div class="col-md-4">
            <div class="form-group">
                <label for="nightly">Default Nightly Rate</label>
                <div class="input-group">
                    <span class="input-group-addon">$</span>
                    <input id="nightly" name="nightly" class="form-control" value="60">
                </div>
            </div>
        </div>

        <div class="col-md-4">
            <div class="form-group">
                <label for="weekly">Default Weekly Rate</label>
                <div class="input-group">
                    <span class="input-group-addon">$</span>
                    <input id="weekly" name="weekly" class="form-control">
                </div>
            </div>
        </div>

        <div class="col-md-4">
            <div class="form-group">
                <label for="monthly">Default Monthly Rate</label>
                <div class="input-group">
                    <span class="input-group-addon">$</span>
                    <input id="monthly" name="monthly" class="form-control">
                </div>
            </div>
        </div>
    </div>
</div>
```

## Validating `sometimes` + `required_without*` + `numeric`

Validation rules for this form will check at least one field has been entered with an accepted amount.

`TrimStrings` and `ConvertEmptyStringsToNull` middleware will cause two of those inputs to become `null` in the request payload so `FormRequest@rules()` or `$this->validate(request())` in the controller would look something like:

```php
Validator::make([
    'weekly' => '60.00',
    'nightly' => null,
    'monthly' => null,
], [
    'weekly' => 'sometimes|required_without_all:nightly,monthly|numeric|min:50|max:99999999',
    'nightly' => 'sometimes|required_without_all:weekly,monthly|numeric|min:50|max:99999999',
    'monthly' => 'sometimes|required_without_all:weekly,nightly|numeric|min:50|max:99999999',
])->passes();

// false, should be true
```

Here are the errors that are instead generated:

```php
$validator->passes();
// false

$validator->errors()->all();
// [
//     "The nightly must be a number.",
//     "The nightly must be at least 50.",
//     "The monthly must be a number.",
//     "The monthly must be at least 50.",
// ]
```

Before the `nullable` rule was [brought into 5.3](https://github.com/laravel/framework/pull/13538/files) to fix validation of blank primitive values, this set of rules handled this form request in Laravel 5.2:

```php
Validator::make([
    'weekly' => '60.00',
    'nightly' => null,
    'monthly' => null,
], [
    'weekly' => 'required_without_all:nightly,monthly|numeric|min:50|max:99999999',
    'nightly' => 'required_without_all:week,monthly|numeric|min:50|max:99999999',
    'monthly' => 'required_without_all:week,nightly|numeric|min:50|max:99999999'
])->passes();

// true
```

## Symptoms of `nullable` rule changes

I know in 5.3+ it's failing because `sometimes` only checks if the _key_ exists in the validator/request payload. The typical use cases of `sometimes` in a POST submit form would be:

* a checkbox isn't ticked (so that field isn't submitted)
* JavaScript disables an optional input element

In this case, the input is enabled and the key exists on `request()->all()` but `numeric|min|max` rules shouldn't be run on the input when its value isn't _present_, in this case `null`.

Replacing `sometimes` within `nullable` will cause `required_without_all` to act as a pass-thru rule that's never applied on empty input.

So should `sometimes` also check for `null` values? Or should another `sometimes_nullable` "flag" validation rule be added for `Validator@isValidatable()` to conditionally apply primitive value rules?

## Another case including a PHPUnit test

Here's another example that didn't work as expected:

```php
public function testValidateSometimesRequiredWithoutIgnoresPrimitiveValueRules()
{
    $v = Validator::make([
        'amount' => '50',
        'percentage' => null,
    ], [
        'amount' => 'sometimes|required_without:percentage|numeric|min:1|max:99999999',
        'percentage' => 'sometimes|required_without:amount|numeric|between:0,1',
    ]);

    $this->assertTrue($v->passes());
}

// instead
$v->passes(); // false
$v->errors()->all(); // ["The percentage must be a number.",]
```

## `$validator->sometimes()` callback in `FormRequest`

My secondary question here would be conditionally applying a `sometimes` callback in a `FormRequest` class as I don't see a reusable [documented way](http://themsaid.com/laravel-advanced-validation-conditionally-adding-rules-20170110/). Studying `FormRequest` internals, this might be a solution?

```php
public function StoreRateRequest
{
    // ...

    public function rules()
    {
        return [
            'weekly' => 'required_without_all:nightly,monthly',
            'nightly' => 'required_without_all:week,monthly',
            'monthly' => 'required_without_all:week,nightly'
        ];
    }

    public function withValidator($validator)
    {
        foreach (['weekly', 'nightly', 'monthly'] as $key) {
            $validator->sometimes($key, 'numeric|min:1|max:99999999', function ($data) use ($key) {
                return !is_null($data->$key);
            });
        }
    }
}
```

Migrating a site from 5.2 to 5.4, I've come across at least a dozen instances of this same problem for any field with a conditional `required_*` rule + any primitive value rule.

> required_without:foo,bar,...
> 
> The field under validation must be present and not empty only when any of the other specified fields are not present.
> 
> required_without_all:foo,bar,...
> 
> The field under validation must be present and not empty only when all of the other specified fields are not present.

I understand the `sometimes()` callback for data-specific business domain choices (like Mohamed's blog post showcased) but it looks like  the `nullable` changes now make `required_*` rules more difficult to manage. Really those rules are now only OK for string data.

Do I just not understand there's a simpler solution available?

**Comment by reganjohnson** [comment]
+1

**Comment by derekmd** [comment]
https://github.com/laravel/framework/pull/18376 committed to 5.4 today might fix this problem. I'll follow-up after verifying so.

It'll be tough luck to apps still on 5.3 that do their own empty string input transforms.

**Comment by derekmd** [comment]
Yup, `nullable` mixed with implicit rules (like `required_without`) _and_ primitive value rules (like `integer`) are now fixed in 5.4.

If this doesn't get backported to Laravel 5.3, this will jerry rig the fix into your app:

## App\Validation\NullableFixValidator.php

```php
namespace App\Validation;

use Illuminate\Support\Arr;
use Illuminate\Validation\Validator as BaseValidator;

class NullableFixValidator extends BaseValidator
{
    protected function isValidatable($rule, $attribute, $value)
    {
        return $this->presentOrRuleIsImplicit($rule, $attribute, $value) &&
            $this->passesOptionalCheck($attribute) &&
            $this->isNotNullIfMarkedAsNullable($rule, $attribute) &&
            $this->hasNotFailedPreviousRuleIfPresenceRule($rule, $attribute);
    }

    protected function isNotNullIfMarkedAsNullable($rule, $attribute)
    {
        if (in_array($rule, $this->implicitRules) || ! $this->hasRule($attribute, ['Nullable'])) {
            return true;
        }

        return ! is_null(Arr::get($this->data, $attribute, 0));
    }
}
```

## App\Providers\ValidatorServiceProvider.php

```php
namespace App\Providers;

use App\Validation\NullableFixValidator;
use Illuminate\Support\ServiceProvider;

class ValidatorServiceProvider extends ServiceProvider
{
    public function boot()
    {
        $this->app['validator']->resolver(function () {
            return new NullableFixValidator(...func_get_args());
        });
    }

    public function register()
    {
    }
}
```

## config/app.php

```php
$appSettings = [
    // ...

    'providers' => [
        // ...
        App\Providers\ValidatorServiceProvider::class,
    ],
];
```
